### ISSUE TITLE
How fast is umongo?

### ISSUE THREAD

**Issue by warvariuc** [Scille/umongo#126]
MongoEngine is slow for creating instances of document models. Also it's slow when accessing fields, because they are descriptors with a lot of logic behind them.
How fast is umongo in these cases?

**Comment by touilleMan** [comment]
Mongoengine is slow for multiples reasons, most of them historic an really hard to fix:
- it does an ensure_index call before every modification in database to guarantee your indexes are in place. This used to be fast because `ensure_index` contains an internal cache, but now pymongo has removed this api (precisely because the internal cache created potential inconsistencies) so now monogengine does a real call to the database to make sure the indexes are in place at every modification.
- referenced document are dereferenced automatically when they are created (e.g. `book.user.pk` would fetch in database to create the `User`, even if we are accessing it primary key which is already known). This makes things extremely slow when iterating over a collection containing references (each document will fetch multiple referenced document one by one instead of doing a batch)
- datastructures are cumbersome with a lot of esoteric which makes them hard to optimise

That's part of the reasons that motivated to create umongo:
- umongo force you to handle the index creation by yourself (so you can choose when you do it knowing you won't have concurrency on it)
- reference field must be explicitly dereferenced with `Reference.fetch` method.
- datastructures are much cleaner to work with

Long story short umongo is faster than mongoengine, especially for real usecases.
On the other hand umongo is obviously slower than raw pymongo, but it shouldn't be a trouble otherwise you wouldn't be using python in the first place ;-)

If you have some time, I encourage you to run your own benchmark of umongo vs mongoengine vs pymongo, it could be really interesting for the project.

**Comment by mindojo-victor** [comment]
We have fixed in our code both index creation and automatic dereference. 

So the main bottleneck we have now is the `__get__` method of fields and simulation of dicts/lists to determine which data was changed.

Does umongo do the same?

**Comment by touilleMan** [comment]
@mindojo-victor, in umongo, fields's values are validated when you set them to your document (or when you load your data from the database).
The good thing about this is 1) your document cannot be in an inconsistant state and 2) data retrieval is very fast because there is no need for complex cooking at this time:

https://github.com/Scille/umongo/blob/90368d7a68e2f149923e5f277a0b43d0edc74a0a/umongo/document.py#L262-L264

On the opposite Mongoengine's `__get__` does a lot of magic which has a high cost (both in term of performances as you found, but also in term of code complexity and bugs...)

> simulation of dicts/lists to determine which data was changed.

Not sure what you talk about, in umongo modified fields are kept trace of with a simple python set. If you have time, I would recommand you to read the code of the [BaseDataProxy class](https://github.com/Scille/umongo/blob/master/umongo/data_proxy.py) which is the one dealing with the data inside each document (so it here you will find the best answer to your questions).

On top of that my advice for you is to try benchmarking umongo against mongoengine, this is the only way to give precise answers to your questions ;-)
